File: cadjoint/viewer/_export.py
"""``POST /api/export``: one object of the program, as a file.

The playground already extracts every scene it shows — dual contouring for
the mesh overlay, the derived B-rep for its sharp edges, a hex or tet mesh
for a study — and :mod:`cadjoint.meshing.export`, :mod:`cadjoint.brep.step`
and :meth:`cadjoint.fem.result.SimulationResult.to_vtk` can each already
write what those passes produce.  This module is the seam between them: a
request names an object and a :class:`~cadjoint.enums.ExportFormat`, the
worker extracts and writes it, and the response is the file.

The two halves live together because they share the vocabulary — which
format takes which options, how the file is named, what content type it
travels under:

- **In the server** (:func:`export_source`): validate the request against
  :class:`~cadjoint.viewer.schema.requests.ExportRequest`, run the worker
  with a temporary path to write to, and hand the bytes back with their
  filename and content type.  The route in :mod:`cadjoint.viewer.playground`
  registers the run as an ``export`` job, so the Processes window sees it
  and can cancel it, and sends the bytes as an attachment.
- **In the worker** (:func:`export_scene`): execute the program, resolve
  the named object, extract on a lattice sized to the object, and write.

**Which STEP writer.**  Two exist.  :func:`cadjoint.meshing.export.save_step`
facets a dual-contour mesh — it recovers flat regions by merging coplanar
quads under an angle threshold and writes a ``PLANE`` per region, with
every curved surface left as triangles.  :func:`cadjoint.brep.step.save_brep_step`
starts from the ownership graph instead: a face there *is* one patch's
zero set, so a plane is exact with its boundary loops (holes included)
collapsed onto the real intersection curves, a full cylindrical band is a
``CYLINDRICAL_SURFACE`` with ``CIRCLE`` edges, and only what the graph
cannot certify (blend faces) is faceted.  The graph writer's output reads
back into OCCT as one valid closed solid with the analytic volume
(``tests/brep/test_step_kernel.py``); the mesh writer's is approximate by
construction.  The export takes the graph path, on the same
:func:`~cadjoint.brep.extract_brep` pass the viewer's edge overlay runs on
every scene, and falls back to the faceted writer when the graph cannot be
derived for an object — so a STEP file is always produced, and the report
says which path produced it.  ``analytic=False`` asks for the faceted
writer outright, which is the comparison baseline the tests use.
"""

from __future__ import annotations

import contextlib
import io
import tempfile
from pathlib import Path
from typing import Any

from pydantic import ValidationError

from cadjoint.enums import ExportFormat, listed
from cadjoint.viewer._limits import OVERSIZED_SOURCE_ERROR, exceeds_source_limit
from cadjoint.viewer.schema.requests import ExportRequest

__all__ = [
    "EXPORT_CONTENT_TYPES",
    "EXPORT_EXTENSIONS",
    "EXPORT_TIMEOUT_SECONDS",
    "export_filename",
    "export_scene",
    "export_source",
    "validate_export_request",
]

# A geometry export is one lattice sweep plus, for STEP, the graph
# extraction the mesh overlay pays for every scene (~17 s warm on the
# starter at the overlay's resolution, and the request may ask for finer).
# A VTK export is a full solve, so it takes the simulate budget instead
# (see :func:`export_source`).  Either is visible and cancellable as a job.
EXPORT_TIMEOUT_SECONDS = 300

#: The file extension each format is written under.
EXPORT_EXTENSIONS: dict[ExportFormat, str] = {
    ExportFormat.OBJ: "obj",
    ExportFormat.STL: "stl",
    ExportFormat.STEP: "step",
    ExportFormat.VTK: "vtk",
}

#: What the response is served as.  OBJ and STL have IANA ``model/`` types;
#: STEP's registered type is ``model/step``, which browsers do not know, so
#: the download is what matters and ``application/step`` is what CAD tools
#: send.  Legacy VTK has no registered type at all.
EXPORT_CONTENT_TYPES: dict[ExportFormat, str] = {
    ExportFormat.OBJ: "model/obj",
    ExportFormat.STL: "model/stl",
    ExportFormat.STEP: "application/step",
    ExportFormat.VTK: "application/octet-stream",
}


def export_filename(name: str, fmt: ExportFormat) -> str:
    """``scene.stl`` — the object's name under the format's extension.

    The name is a Python identifier (or a study name), so the only thing to
    guard is a name that is not a safe file name at all.
    """
    stem = "".join(char if (char.isalnum() or char in "-_.") else "_" for char in name.strip())
    return f"{stem.strip('.') or 'export'}.{EXPORT_EXTENSIONS[fmt]}"


def _validation_message(error: ValidationError) -> str:
    """The first field failure, in the dialog's own words."""
    first = error.errors()[0]
    location = ".".join(str(part) for part in first.get("loc", ())) or "request"
    if location == "format":
        return f"Export `format` must be one of: {listed(ExportFormat)}."
    return f"Export `{location}`: {first.get('msg', 'is invalid')}."


def validate_export_request(
    request: dict[str, Any],
) -> tuple[dict[str, Any] | None, ExportRequest | None]:
    """``(error, parsed)`` — the request against its model, before any work.

    Args:
        request: The raw request object.

    Returns:
        ``(None, model)`` for a valid request, or ``({"ok": False, ...},
        None)`` with the message the dialog shows.
    """
    try:
        parsed = ExportRequest.model_validate(request)
    except ValidationError as error:
        return {"ok": False, "error": _validation_message(error)}, None
    if exceeds_source_limit(parsed.source):
        return {"ok": False, "error": OVERSIZED_SOURCE_ERROR}, None
    return None, parsed


# ── the server half ─────────────────────────────────────────────────────────


def export_source(request: dict[str, Any], timeout: float | None = None) -> dict[str, Any]:
    """Validate an export request and write the file in a disposable child process.

    The worker writes to a temporary path this process owns; the bytes are
    read back and the directory removed before returning, so nothing is
    left on disk and nothing large crosses the worker's JSON pipe.

    Args:
        request: The raw request body.
        timeout: The worker budget; by default :data:`EXPORT_TIMEOUT_SECONDS`,
            or the simulate budget for ``vtk`` (which solves the study).

    Returns:
        On success ``{"ok": True, "format", "name", "filename",
        "content_type", "size", "data": bytes, "report", "output"}``; on
        failure the worker's ``{"ok": False, "error"}`` (with
        ``error_kind="fem_unavailable"`` when a VTK export has no solver).
    """
    from cadjoint.viewer._worker_client import SIMULATE_TIMEOUT_SECONDS, _run_worker

    error, parsed = validate_export_request(request)
    if error is not None or parsed is None:
        return error or {"ok": False, "error": "Invalid export request."}
    if timeout is None:
        timeout = (
            SIMULATE_TIMEOUT_SECONDS
            if parsed.format is ExportFormat.VTK
            else EXPORT_TIMEOUT_SECONDS
        )
    filename = export_filename(parsed.name, parsed.format)
    with tempfile.TemporaryDirectory(prefix="cadjoint-export-") as folder:
        path = Path(folder) / filename
        extra = parsed.model_dump(mode="json", exclude={"source"})
        extra["path"] = str(path)
        result = _run_worker(parsed.source, "export", timeout, extra=extra)
        if not result.get("ok"):
            return result
        if not path.is_file():
            return {"ok": False, "error": "The export worker finished without writing a file."}
        data = path.read_bytes()
    return {
        "ok": True,
        "format": parsed.format.value,
        "name": parsed.name,
        "filename": filename,
        "content_type": EXPORT_CONTENT_TYPES[parsed.format],
        "size": len(data),
        "data": data,
        "report": result.get("report") or {},
        "output": result.get("output", ""),
    }


# ── the worker half ─────────────────────────────────────────────────────────


def _exportable_names(namespace: dict[str, Any]) -> list[str]:
    """Every variable of the program bound to an SDF, ``scene`` first."""
    from cadjoint.sdf.base import SDF

    names = [
        name
        for name, value in namespace.items()
        if not name.startswith("_") and isinstance(value, SDF)
    ]
    return sorted(names, key=lambda name: (name != "scene", name))


def _named_object(namespace: dict[str, Any], name: str) -> Any:
    """The SDF the program bound to *name*, or raise listing the alternatives."""
    from cadjoint.sdf.base import SDF

    value = namespace.get(name)
    if isinstance(value, SDF) and not name.startswith("_"):
        return value
    exportable = ", ".join(repr(candidate) for candidate in _exportable_names(namespace))
    if value is None:
        raise ValueError(
            f"The program binds no SDF object named {name!r} (exportable: {exportable or 'none'})."
        )
    raise ValueError(
        f"{name!r} is a {type(value).__name__}, not an SDF object "
        f"(exportable: {exportable or 'none'})."
    )


def _object_grid(sdf: Any, resolution: int) -> Any:
    """A uniform lattice fitted to the object's inside region.

    The bounds come from the same coarse scan a ``SimMesh`` without explicit
    bounds uses, so an exported part and its simulation mesh agree on where
    the part is.  ``resolution`` counts cells along the longest axis and
    the other two axes take proportionally fewer, which keeps the spacing
    isotropic — a cube of cells over a flat plate would spend most of its
    budget on the thin axis.
    """
    import numpy as np

    from cadjoint.fem.simmesh import _scan_bounds
    from cadjoint.meshing import GridSpec

    bounds, size = _scan_bounds(sdf, padding=0.0)
    extent = np.asarray(size, dtype=np.float64)
    cells = tuple(int(max(1, round(resolution * axis / float(extent.max())))) for axis in extent)
    return GridSpec.from_bounds(bounds, size, cells)


def _write_geometry(obj: Any, request: ExportRequest, path: Path) -> dict[str, Any]:
    """Extract one SDF object and write it as OBJ, STL or STEP."""
    import warnings

    import jax.numpy as jnp

    from cadjoint.meshing.dual_contouring import extract_mesh
    from cadjoint.meshing.export import save_obj, save_step, save_stl

    sdf = lambda p: jnp.asarray(obj(p))  # noqa: E731
    grid = _object_grid(sdf, request.resolution)
    report: dict[str, Any] = {
        "resolution": list(grid.cells),
        "bounds": list(grid.origin),
        "size": [spacing * count for spacing, count in zip(grid.spacing, grid.cells)],
    }

    if request.format is ExportFormat.STEP and request.analytic:
        # The graph path: exact planes and cylinders, faceted where the graph
        # cannot certify a face.  See the module docstring for why this
        # writer and not the mesh one.
        from cadjoint.brep import extract_brep, save_brep_step

        try:
            brep = extract_brep(obj, grid)
            written = save_brep_step(brep, path)
        except Exception as error:  # noqa: BLE001 - the faceted writer is the fallback
            report["fallback"] = f"{type(error).__name__}: {error}"[:500]
        else:
            report["path"] = "brep"
            report["faces"] = written.get("faces", {})
            return report

    with warnings.catch_warnings():
        # The lattice is fitted to the object with a margin, so the open
        # boundary warning can only mean a scene that fills the default
        # scan volume (a ground plane); clipping it is the point.
        warnings.filterwarnings("ignore", message="The isosurface crosses the extraction boundary")
        mesh = extract_mesh(sdf, grid)
    report["vertices"] = int(mesh.vertices.shape[0])
    report["triangles"] = int(mesh.faces.shape[0])
    if request.format is ExportFormat.OBJ:
        save_obj(mesh, path, merge_planar=request.merge_planar)
    elif request.format is ExportFormat.STL:
        save_stl(mesh, path, binary=request.binary)
    else:
        save_step(mesh, path)
        report["path"] = "mesh"
    return report


def _write_result(
    scene: Any, studies: list[Any], request: ExportRequest, path: Path
) -> dict[str, Any]:
    """Solve (or reuse) the named study and write its fields as VTK."""
    import jax.numpy as jnp

    from cadjoint.viewer._worker_scene import _FEM_UNAVAILABLE_MESSAGE, _named_study

    study = _named_study(studies, request.name)
    try:
        import jax_fem  # noqa: F401
    except ImportError:
        return {"ok": False, "error_kind": "fem_unavailable", "error": _FEM_UNAVAILABLE_MESSAGE}
    sdf = lambda p: jnp.asarray(scene(p))  # noqa: E731
    # A result the program computed itself (a module-level solve) is the
    # result the user is looking at; only solve when there is none.
    cached = study.last_result is not None
    result = study.last_result if cached else study.solve(sdf)
    result.to_vtk(str(path))
    return {"ok": True, "report": {"study": study.name, "cached": cached}}


def export_scene(request: dict[str, Any]) -> dict[str, Any]:
    """The worker's ``mode="export"``: run the program, write one file.

    ``request["path"]`` is where the server wants the file; everything else
    is an :class:`ExportRequest` field.  Returns ``{"ok": True, "report",
    "output"}`` — the bytes stay on disk for the server to collect.
    """
    # The worker protocol adds `mode` and this module adds `path`; neither
    # is a field of the request the model describes.
    fields = {key: value for key, value in request.items() if key not in ("mode", "path")}
    error, parsed = validate_export_request(fields)
    if error is not None or parsed is None:
        return error or {"ok": False, "error": "Invalid export request."}
    path = Path(str(request.get("path") or ""))
    if not path.name:
        return {"ok": False, "error": "The export worker was given no path to write to."}

    from cadjoint.viewer._worker_scene import _execute_scene

    captured = io.StringIO()
    with contextlib.redirect_stdout(captured), contextlib.redirect_stderr(captured):
        namespace = _execute_scene(parsed.source)
        if parsed.format is ExportFormat.VTK:
            result = _write_result(namespace["scene"], namespace["__studies__"], parsed, path)
        else:
            report = _write_geometry(_named_object(namespace, parsed.name), parsed, path)
            result = {"ok": True, "report": report}
    result["output"] = captured.getvalue()[-8_000:]
    return result

File: frontend/src/export.ts
/**
 * File → Export…, the part with no DOM in it.
 *
 * The dialog is a thin form over `POST /api/export`; everything it has to
 * *decide* — which formats there are and what each one takes, which names in
 * the program are worth offering, how a request body is shaped, what a
 * download is called — is here, as plain functions a unit test can call.
 * `api.ts` does the fetch and `components/ExportDialog.tsx` draws the form;
 * neither computes anything this module does not.
 */

import type { ExportFormat, ExportRequest } from "./types";

/**
 * The lattice budget, mirrored from `cadjoint/viewer/_limits.py`.
 *
 * The server refuses a request outside the bracket with a message the dialog
 * shows, so these only have to keep the control honest — a number box whose
 * range says the same thing the validator does.
 */
export const EXPORT_RESOLUTION = { min: 8, max: 256, default: 64 } as const;

/** What a format takes: an SDF variable of the program, or a declared study. */
export type ExportTarget = "object" | "study";

export interface ExportFormatInfo {
  value: ExportFormat;
  /** The label on the picker — the format's own name, as CAD tools write it. */
  label: string;
  extension: string;
  takes: ExportTarget;
  /** One sentence under the picker saying what the file will contain. */
  note: string;
  /** The one option the format has, if it has one. */
  option: { key: "binary" | "analytic" | "merge_planar"; label: string } | null;
}

/**
 * Every format the server writes, in the order the picker lists them.
 *
 * STEP's note says what makes it the interesting one: it is not a faceted
 * mesh with a `.step` extension but the derived B-rep — exact planes and
 * cylinders with their real boundary curves — faceted only where a face is
 * a blend the graph cannot certify (`cadjoint/viewer/_export.py`).
 */
export const EXPORT_FORMATS: readonly ExportFormatInfo[] = [
  {
    value: "stl",
    label: "STL",
    extension: "stl",
    takes: "object",
    note: "Triangles of the dual-contour surface. Binary by default; ASCII on request.",
    option: { key: "binary", label: "Binary" },
  },
  {
    value: "obj",
    label: "OBJ",
    extension: "obj",
    takes: "object",
    note: "Wavefront OBJ. Flat regions become single n-gon faces; curved regions stay triangles.",
    option: { key: "merge_planar", label: "Merge planar faces" },
  },
  {
    value: "step",
    label: "STEP",
    extension: "step",
    takes: "object",
    note: "AP214 from the derived B-rep: exact planes and cylinders with their real edge curves; blends faceted.",
    option: { key: "analytic", label: "Analytic surfaces" },
  },
  {
    value: "vtk",
    label: "VTK",
    extension: "vtk",
    takes: "study",
    note: "A declared study's solved mesh and fields, for ParaView. Solves the study if the program has not.",
    option: null,
  },
];

/** The picker entry for a format. */
export function formatInfo(format: ExportFormat): ExportFormatInfo {
  const info = EXPORT_FORMATS.find((entry) => entry.value === format);
  if (!info) throw new Error(`Unknown export format ${format}.`);
  return info;
}

/**
 * Constructors whose result is never an SDF, so a variable bound to one is
 * not worth offering as an export target. Everything else — a primitive, a
 * boolean, a transform, an extrusion, a `Solid.box(...)` — might be.
 */
const NOT_GEOMETRY =
  /^(?:Scalar|Vector|Vector2|Material|SimMesh|ThermalStudy|ElasticStudy|Optimization|PolygonProfile|SketchPlane|jnp|np|jax)\b/;

/** A right-hand side that is plainly a literal, not a call. */
const LITERAL = /^(?:[-+]?\d|["'[({]|True\b|False\b|None\b)/;

/**
 * The names the program binds at module level that could be geometry.
 *
 * Read off the text, not the payload: the compile payload names features the
 * viewer can outline, and the thing a user most often wants to export — the
 * `Union(...)` a whole assembly is bound to — is not one of those. A
 * left-hand identifier on an unindented line is cheap to find and honest
 * about what it is: a suggestion. The server checks the real object and, on
 * a miss, answers with the names that would have worked.
 *
 * `scene` leads (it is the default and the one that always exists); the rest
 * keep source order. Private names, loop variables and literals are skipped.
 */
export function candidateObjects(source: string): string[] {
  const names: string[] = [];
  for (const line of source.split("\n")) {
    const match = /^([A-Za-z][A-Za-z0-9_]*)\s*(?::\s*[^=]+)?=(?!=)\s*(.*)$/.exec(line);
    if (!match) continue;
    const [, name, rhs] = match;
    if (NOT_GEOMETRY.test(rhs) || LITERAL.test(rhs)) continue;
    if (!names.includes(name)) names.push(name);
  }
  return ["scene", ...names.filter((name) => name !== "scene")];
}

/** What the name picker offers for a format: objects, or declared studies. */
export function exportTargets(
  format: ExportFormat,
  source: string,
  studyNames: readonly string[],
): string[] {
  return formatInfo(format).takes === "study" ? [...studyNames] : candidateObjects(source);
}

/** The name a fresh dialog starts on for a format. */
export function defaultExportName(
  format: ExportFormat,
  studyNames: readonly string[],
): string {
  return formatInfo(format).takes === "study" ? (studyNames[0] ?? "") : "scene";
}

export interface ExportOptions {
  format: ExportFormat;
  name: string;
  resolution: number;
  binary: boolean;
  analytic: boolean;
  mergePlanar: boolean;
}

/** A resolution the server will accept: an integer inside the bracket. */
export function clampResolution(value: number): number {
  if (!Number.isFinite(value)) return EXPORT_RESOLUTION.default;
  return Math.min(EXPORT_RESOLUTION.max, Math.max(EXPORT_RESOLUTION.min, Math.round(value)));
}

/**
 * The request body for one export.
 *
 * Only the option the format actually reads is sent — the server accepts all
 * three on any format, but a `binary: false` riding along on a STEP request
 * is the kind of thing that reads as a bug in a job's field list.
 */
export function exportRequest(source: string, options: ExportOptions): ExportRequest {
  const info = formatInfo(options.format);
  const body: ExportRequest = {
    source,
    format: options.format,
    name: options.name.trim(),
  };
  if (info.takes === "object") body.resolution = clampResolution(options.resolution);
  if (info.option?.key === "binary") body.binary = options.binary;
  if (info.option?.key === "analytic") body.analytic = options.analytic;
  if (info.option?.key === "merge_planar") body.merge_planar = options.mergePlanar;
  return body;
}

/**
 * The file name a `Content-Disposition` header carries, or the fallback.
 *
 * Handles the quoted plain form the server sends and the RFC 5987 `filename*`
 * form a proxy might rewrite it into.
 */
export function attachmentFilename(header: string | null | undefined, fallback: string): string {
  if (!header) return fallback;
  const extended = /filename\*\s*=\s*(?:UTF-8|utf-8)''([^;]+)/.exec(header);
  if (extended) {
    try {
      return decodeURIComponent(extended[1].trim()) || fallback;
    } catch {
      return fallback;
    }
  }
  const quoted = /filename\s*=\s*"([^"]*)"/.exec(header);
  if (quoted) return quoted[1] || fallback;
  const bare = /filename\s*=\s*([^;\s]+)/.exec(header);
  return bare ? bare[1] : fallback;
}

/**
 * The download's name: the scene's stem, then the object's, then the extension.
 *
 * `heatsink-scene.stl` says which program the part came from; `scene.stl`
 * on its own does not, and an unsaved buffer has nothing to say, so it
 * keeps the server's name.
 */
export function downloadName(
  serverFilename: string,
  sceneName: string | null | undefined,
): string {
  if (!sceneName) return serverFilename;
  const stem = sceneName.replace(/\.py$/i, "").trim();
  return stem ? `${stem}-${serverFilename}` : serverFilename;
}

/**
 * A worker failure, in one line.
 *
 * The worker answers an exception with its whole traceback — right for the
 * console, wrong for a dialog. The last non-empty line of a traceback is the
 * exception itself, which is the sentence a user can act on.
 */
export function errorSummary(text: string): string {
  const lines = text
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0);
  if (lines.length === 0) return "Export failed.";
  if (!/^Traceback \(most recent call last\)/.test(lines[0])) return lines.join(" ");
  return lines[lines.length - 1];
}
